velox/expression/CastExpr-inl.h

```diff
@@ -301,6 +325,86 @@ VectorPtr CastExpr::applyDecimalToBooleanCast(
   return result;
 }
 
+template <typename FromNativeType>
+VectorPtr CastExpr::applyDecimalToVarcharCast(
+    const SelectivityVector& rows,
+    const BaseVector& input,
+    exec::EvalCtx& context,
+    const TypePtr& fromType) {
+  VectorPtr result;
+  context.ensureWritable(rows, VARCHAR(), result);
+  (*result).clearNulls(rows);
+  const auto simpleInput = input.as<SimpleVector<FromNativeType>>();
+  int precision = getDecimalPrecisionScale(*fromType).first;
+  int scale = getDecimalPrecisionScale(*fromType).second;
+  // A varchar's size is estimated with unscaled value digits, dot, leading
+  // zero, and possible minus sign.
+  int varcharSize = precision + !!(scale > 0) + !!(precision == scale) + 1;
+  // Calculate the max total size of the raw string buffer.
+  size_t maxTotalResultBytes = rows.countSelected() * varcharSize;
+
+  auto flatResult = result->asFlatVector<StringView>();
+  Buffer* buffer = flatResult->getBufferWithSpace(maxTotalResultBytes);
+  size_t bufferSize = buffer->size();
+  char* rawBuffer = buffer->asMutable<char>() + bufferSize;
+  buffer->setSize(bufferSize + maxTotalResultBytes);
+  size_t offset = 0;
+  applyToSelectedNoThrowLocal(context, rows, result, [&](vector_size_t row) {
+    if (simpleInput->isNullAt(row)) {
+      result->setNull(row, true);
+    } else {
+      auto unscaledValue = simpleInput->valueAt(row);
+      const char zero = '0';
+      size_t startOffset = offset;
+      if (unscaledValue == 0) {
+        memcpy(rawBuffer + offset, &zero, sizeof(char));
+        offset += sizeof(char);
+      } else {
+        if (unscaledValue < 0) {
+          const char minusSign = '-';
+          memcpy(rawBuffer + offset, &minusSign, sizeof(char));
+          offset += sizeof(char);
+          unscaledValue = ~unscaledValue + 1;
+        }
+        auto [position, ec] = std::to_chars(
+            rawBuffer + offset,
+            rawBuffer + offset + varcharSize,
+            unscaledValue / DecimalUtil::kPowersOfTen[scale]);
+        VELOX_USER_CHECK(
+            ec == std::errc(),
+            "Cast from decimal to varchar fails because {}.",
+            std::make_error_code(ec).message());
+        offset = position - rawBuffer;
+        if (scale > 0) {
+          const char dot = '.';
+          memcpy(rawBuffer + offset, &dot, sizeof(char));
+          offset += sizeof(char);
+
+          uint128_t fraction = unscaledValue % DecimalUtil::kPowersOfTen[scale];
+          // Append leading zeros.
+          for (size_t i = 0; i < std::max(scale - countDigits(fraction), 0);
```

Comment:
compute the for loop times first.
auto digitsNum = std::max(scale - countDigits(fraction), 0);

Reply:
Updated.